### PR TITLE
Jmaron/vz 3509

### PR DESCRIPTION
**Pull request by jmaron99**
# Description

- Jenkinsfile that executes release and patch steps up to the manual publication of images to OCR
- Script for release branch creation
- Script that serves as a credential provider for git, leveraging the r/w git credential configured in Jenkins.  Supports the need to push the release branch created.

Note that actual final testing of this pipeline requires the Jenkinsfile to be available in the master branch.  

OCR publication is disabled if the TEST_RUN input is checked (defaults to false)

Pre-release validation checks are disabled if IGNORE_PRE_RELEASE_VALIDATION_FAILURES is checked (defaults to false).

Fixes VZ-3509

# Checklist 

As the author of this PR, I have:

- [ ] Checked that I included or updated copyright and license notices in all files that I altered
- [ ] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: release/scripts/create_branch.sh [hunk 1]
```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env bash
+#
+# Copyright (c) 2021, Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
+#
+
+function usage {
+    echo
+    echo "usage: $0 [-v release_version] [-c source_commit]"
+    echo "  -v release_version         The release version (e.g. #.# for a major release, #.#.# for a patch release. Required"
```
Comment (fredtibbitts) [inline release/scripts/create_branch.sh:10]: Minor, help text needs updating, version is always major.minor.patch

File: release/builds/Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+            registryUrl "${RUNNER_DOCKER_REGISTRY_URL}"
+            registryCredentialsId 'ocir-pull-and-push-account'
+            label "internal"
+        }
+    }
+
+    parameters {
+        string (description: 'The source branch for the release', name: 'SOURCE_BRANCH', defaultValue: 'master', trim: true)
+        string (description: 'The source commit for the release (required for full release)', name: 'COMMIT_TO_USE', defaultValue: 'NONE', trim: true )
+        string (description: 'The target version (release: #.#, patch: #.#.#)', name: 'TARGET_VERSION', defaultValue: 'NONE', trim: true)
```
Comment (fredtibbitts) [inline release/builds/Jenkinsfile:27]: major.minor.patch

File: release/builds/Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+                            branches: [[name: params.COMMIT_TO_USE]],
+                            doGenerateSubmoduleConfigurations: false,
+                            extensions: scm.extensions + [[$class: 'LocalBranch'], [$class: 'WipeWorkspace']],
+                            submoduleCfg: [],
+                            userRemoteConfigs: [[url: env.SCM_VERRAZZANO_GIT_URL, credentialsId:'github-markxnelns-private-access-token']]])
+                        env.GIT_COMMIT = scmInfo.GIT_COMMIT
+                        env.GIT_BRANCH = scmInfo.GIT_BRANCH
+                        // If the commit we were handed is not what the SCM says we are using, fail
+                        if (!env.GIT_COMMIT.equals(params.COMMIT_TO_USE)) {
+                            echo "SCM didn't checkout the commit we expected. Expected: ${params.GIT_COMMIT_TO_USE}, Found: ${scmInfo.GIT_COMMIT}"
```
Comment (fredtibbitts) [inline release/builds/Jenkinsfile:69]: I had copied this same block of code from another Jenkinsfile and tried to use it, but I had problems, I don't think you can set `env.GIT_COMMIT` and `env.GIT_BRANCH`, if I recall correctly? I think those are set by the system and can't be changed? Also the echoed message doesn't reflect the actual check, so if you do leave this in you should probably change `scmInfo.GIT_COMMIT` to `env.GIT_COMMIT`.

File: release/builds/Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+                            if (env.IS_PATCH_RELEASE == 'true') {
+                                RELEASE_JOB_NAME = 'verrazzano/' + env.SOURCE_BRANCH
+                            } else {
+                                RELEASE_JOB_NAME = 'verrazzano/release-' + params.TARGET_VERSION
+                            }
+                            // if branch build not found, trigger repo scan and wait
+                            if (Jenkins.instance.getItemByFullName(RELEASE_JOB_NAME) == null) {
+                                Jenkins.instance.getItemByFullName('verrazzano').scheduleBuild()
+                                while (Jenkins.instance.getItemByFullName(RELEASE_JOB_NAME) == null) {
+                                    sleep(1)
```
Comment (fredtibbitts) [inline release/builds/Jenkinsfile:154]: Maybe echo something here so we can see in the console output if this gets stuck looping for some reason and we have to kill it?

File: release/builds/Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+        stage('Pre-release validation') {
+            environment {
+                IGNORE_FAILURES = "${params.IGNORE_PRE_RELEASE_VALIDATION_FAILURES}"
+            }
+            steps {
+                script {
+                    sh """
+                        cd ${WORKSPACE}
+                        ./release/scripts/prerelease_validation.sh ${params.TARGET_VERSION}
```
Comment (fredtibbitts) [inline release/builds/Jenkinsfile:118]: One thing I'm curious about, the scm checkout does a very limited checkout and the prerelease validation script depends on getting the latest version tag from the repo. Do you know if all of the tags are being fetched? We could do a "git tag" after the prerelease_validation.sh script is run as a test just to make sure the tags are really being fetched.

**Review by fredtibbitts — COMMENTED** [review]
LGTM, just a few minor things.

File: release/builds/Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+                }
+
+                stage('Verrazzano Build') {
+                    steps {
+                        script {
+                            if (env.IS_PATCH_RELEASE == 'true') {
+                                RELEASE_JOB_NAME = 'verrazzano/' + env.SOURCE_BRANCH
+                            } else {
+                                RELEASE_JOB_NAME = 'verrazzano/release-' + params.TARGET_VERSION
```
Comment (agv426) [inline release/builds/Jenkinsfile:148]: If I'm reading correctly for patch releases I think we will need a separate variable that has the minor version so we can build on that branch.

Ie: patch 1.0.n would be building on release-1.0

If TARGET_VERSION is 1.0.1 then we will need to trim the patch level off to get the minor verion

File: release/builds/Jenkinsfile (hunk 5 above)
Comment (jmaron99) [reply release/builds/Jenkinsfile:118]: In my testing I discovered that it did compare the to the 1.0.1 tag which I assumed, by date, was the latest tag

File: release/builds/Jenkinsfile (hunk 6 above)
Comment (jmaron99) [reply release/builds/Jenkinsfile:148]: Yep - missed that when I made the changes this morning.  will do.

File: release/builds/Jenkinsfile (hunk 4 above)
Comment (jmaron99) [reply release/builds/Jenkinsfile:154]: the sleep command echoes out to the log so there is an indication each time the sleep is invoked

Comment (fredtibbitts) [reply release/builds/Jenkinsfile:155]: ah, right.

File: release/builds/Jenkinsfile (hunk 5 above)
Comment (fredtibbitts) [reply release/builds/Jenkinsfile:118]: Great, thanks for confirming.

**Review by fredtibbitts — APPROVED** [review]

**Pull request merged**